src/chains/filecoin/filecoin/src/blockchain.ts

```diff
@@ -188,11 +191,39 @@ export default class Blockchain extends Emittery.Typed<
   }
 
   private async getIPFSObjectSize(cid: string): Promise<number> {
-    let stat = await this.ipfsServer.node.object.stat(cid, {
+    let stat = await this.ipfsServer.node.object.stat(cid as any, {
       timeout: 500 // Enforce a timeout; otherwise will hang if CID not found
     });
 
-    return stat.DataSize;
+    return stat.CumulativeSize;
+  }
+
+  private async downloadFile(cid: string, ref: FileRef): Promise<void> {
+    const size = await this.getIPFSObjectSize(cid);
+    const content = new Uint8Array(size);
+    const chunks = this.ipfsServer.node.files.read(new IPFS_CID(cid), {
+      timeout: 500 // Enforce a timeout; otherwise will hang if CID not found
+    });
+
+    let index = 0;
+    for await (const chunk of chunks) {
+      content.set(chunk, index);
+      index += chunk.byteLength;
+    }
+
+    const dirname = path.dirname(ref.path);
+    try {
+      if (!fs.existsSync(dirname)) {
+        await fs.promises.mkdir(dirname, { recursive: true });
+      }
+      await fs.promises.writeFile(ref.path, content.slice(0, index), "binary");
```

Comment:
Disregard this comment if you refactor this to pipe the chunks directly to disk...why do we need `content.slice(0, index)`, here?

Reply:
I will write directly to disk, but FWIW, the slice here is because we allocated more RAM in `content` than we needed. since `ipfs.object.stat().CumulativeSize` includes IPFS overhead of the DAG node storage. For example a 5 byte text file actually requires 13 bytes on IPFS